frontend/ui/widget/inputtext.lua

```diff
@@ -141,8 +143,90 @@ if Device:isTouchDevice() or Device:hasDPad() then
                 self.text_widget:moveCursorToXY(x, y, true) -- restrict_to_view=true
                 self.charpos, self.top_line_num = self.text_widget:getCharPos()
             end
-            if Device:hasClipboard() and Device.input.hasClipboardText() then
-                self:addChars(Device.input.getClipboardText())
+            -- clipboard dialog
+            if Device:hasClipboard() then
+                if self.do_select then -- select mode on
+                    if self.selection_start_pos then -- select end
+                        Device.input.setClipboardText(table.concat(self.charlist, "", self.selection_start_pos, self.charpos-1))
+                        UIManager:show(Notification:new{
+                            text = _("Text copied to clipboard"),
+                        })
+                        self.selection_start_pos = nil
+                        self.do_select = false
+                        return true
+                    else -- select start
+                        self.selection_start_pos = self.charpos
+                        UIManager:show(Notification:new{
+                            text = _("Hold the end of selection"),
+                        })
+                        return true
+                    end
+                end
+                local input_dialog
+                input_dialog = require("ui/widget/inputdialog"):new{
+                    title = _("Clipboard"),
+                    stop_events_propagation = true,
+                    input_hint = _("empty"),
+                    input = Device.input.getClipboardText(),
+                    buttons = {
+                        {
+                            {
+                                text = _("All"),
+                                callback = function()
+                                    UIManager:close(input_dialog)
+                                    Device.input.setClipboardText(table.concat(self.charlist))
+                                end,
+                            },
+                            {
+                                text = _("Line"),
+                                callback = function()
+                                    UIManager:close(input_dialog)
+                                    Device.input.setClipboardText(table.concat(self.charlist, "", self:getStringPos({"\n"}, {"\n"})))
+                                end,
+                            },
+                            {
+                                text = _("Word"),
+                                callback = function()
+                                    UIManager:close(input_dialog)
+                                    Device.input.setClipboardText(table.concat(self.charlist, "", self:getStringPos({"\n", " "}, {"\n", " "})))
```

Comment:
These are a bit long.
You could split that on 2 lines:
```lua
local txt = ...
Device.input.setClipboardText(txt)
```
(You're splitting by spaces, but I don't think we have anything in frontend that check word boundaries or punctuations - so this might be good enough.)

Reply:
Fixed long lines.
I thought of adding more delimiters, but when editing lua, "words" might include `:` or `{` or `/` so on.
So left eol and spaces only.